app/dao/service_broadcast_settings_dao.py

```diff
@@ -0,0 +1,15 @@
+from app import db
+from app.models import ServiceBroadcastSettings
+
+
+def insert_or_update_service_broadcast_settings(service, channel, provider_restriction=None):
```

Comment:
So this is interesting...

Why do we want to commit the transaction immediately after `insert_or_update_service_broadcast_settings` is run? As opposed to waiting for the end of the `set_as_broadcast_service` where everything outstanding gets committed.

I can see that
`dao_remove_service_permission` is not transactional
`dao_add_service_permission` is transactional
`dao_update_service` is transactional
`dao_add_service_to_organisation` is transactional

What makes us decide if a function should be transactional?

I can see some cases for wanting a function to be transactional. When we receive a post request to send an email, we want to save the email to the DB and then put a task on a queue to send the email. It makes sense the have the saving of the email to the DB as transactional because if we put the task on the queue and then attempt to commit the transaction after this and it fails, we are in a dodgy state where we've put a task on a queue for something that doesn't exist.

I don't see a similar case here... but maybe there is a reason still that we want it to be transactional? Maybe it's even just for consistency?

As an interesting aside, I was thinking about the fact that if `dao_remove_service_permission` became transactional then my code would remove all permissions and then would add the permission back for broadcasts. This means if someone already has the broadcast permission it will remove it and then add it again and so you have a few milliseconds without the permission (and whether that would be OK)

Reply:
`dao_remove_service_permission` doesn't have the transactional decorator but it still [commits the session in the function](https://github.com/alphagov/notifications-api/blob/9b940637a536d1716411209674122dce1a54c652/app/dao/service_permissions_dao.py#L21), which does the same thing. I thought we generally followed the pattern of always committing after DAO functions that add or modify data, but maybe there are a few exceptions where we don't want to do that.

If the function is going to be kept without the transactional decorator, it might be worthwhile adding a comment to the function pointing that out (since it's not how many of the other DAO functions work, so it might be easy to miss that there is no database `commit` happening)

Reply:
Ah, I missed the inline commiting on `dao_remove_service_permission`.

That all seems very consistent then and therefore seems like a good case for me to add the transactional decorator. Will do this now

Reply:
Also given this, any thought on the fact that for a few milliseconds, the broadcast permission will have been removed completely, if they already had it, before being added again? I could change my code to not remove the broadcast permission for example. I don't think it's a problem to be honest, the risk of that causing a problem is very very small but thought I'd get another opinion please

Reply:
I've added the transactional decorator and force pushed

Reply:
The question to ask when wrapping db operations with @transactional is would the data be in the right state if the first statement committed but the second statement failed. Would the service make sense if it had the broadcast settings but not the permission?

Reply:
I think in this case all the updates/inserts should be wrapped in a single transaction rather than several separate ones.

Reply:
Thank you Becca, I agree. Will make this change

Reply:
Change pushed up in new commit